hist/hist/inc/THnChain.h

```diff
@@ -0,0 +1,266 @@
+   {
+      fFiles.Add(new TObjString(fileName));
+
+      // Initialize axes from first seen instance.
+      if (fAxes.GetEntries() == 0) {
+         THnBase* hs = ReadHistogram(fileName);
+
+         if (hs) {
+            const Int_t naxes = hs->GetNdimensions();
+            for (Int_t i = 0; i < naxes; ++i) {
+               fAxes[i] = hs->GetAxis(i);
+            }
+         }
+      }
+   }
+
+   /// Get an axis from the histogram.
+   ///
+   /// \param i index of the axis to retrieve
+   ///
+   /// This function requires that a file containing the histogram was
+   /// already added with `AddFile`.
+   ///
+   /// Properties set on the axis returned by `GetAxis` are propagated to all
+   /// histograms in the chain, so it can e.g. be used to set ranges for
+   /// projections.
+   TAxis* GetAxis(Int_t i)
+   {
+      if (i < 0 || i >= fAxes.GetEntries()) {
+         return nullptr;
+      }
+
+      return static_cast<TAxis*>(fAxes[i]);
+   }
+
+   TH1* Projection(Int_t xDim, Option_t* option = "")
+   {
+      // Forwards to `THnBase::Projection()`.
+      Int_t dim[1] = {xDim};
+      return static_cast<TH1*>(ProjectionAny(1, dim, option));
+   }
+
+   TH2* Projection(Int_t yDim, Int_t xDim, Option_t* option = "")
+   {
+      // Forwards to `THnBase::Projection()`.
+      Int_t dim[2] = {xDim, yDim};
+      return static_cast<TH2*>(ProjectionAny(2, dim, option));
+   }
+
+   TH3* Projection(Int_t xDim, Int_t yDim, Int_t zDim, Option_t* option = "")
+   {
+      // Forwards to `THnBase::Projection()`.
+      Int_t dim[3] = {xDim, yDim, zDim};
+      return static_cast<TH3*>(ProjectionAny(3, dim, option));
+   }
+
+   THnBase* ProjectionND(Int_t ndim, const Int_t* dim, Option_t* option = "")
+   {
+      // Forwards to `THnBase::ProjectionND()`.
+      return static_cast<THnBase*>(ProjectionAny(ndim, dim, option));
+   }
+
+ private:
+   TString fName; ///< name of the histogram
+
+   TObjArray fFiles; ///< a list of files to extract the histogram from
+   TObjArray fAxes;  ///< the list of histogram axes
+
+   /// Projects all histograms in the chain.
+   ///
+   /// See `THnBase::Projection` for parameters and their semantics.
+   TObject* ProjectionAny(Int_t ndim, const Int_t* dim, Option_t* option = "")
+   {
+      if (ndim == 0) {
+         return nullptr;
+      }
+
+      TObject* h_merged = nullptr;
+      for (const auto& obj : fFiles) {
+         THnBase* hs = ReadHistogram(static_cast<TObjString*>(obj)->String());
+
+         if (!hs || !CheckConsistency(*hs, fAxes)) {
+            Warning("ProjectionAny",
+                    "Could find histogram matching template %s in file %s",
+                    fName.Data(),
+                    static_cast<TObjString*>(obj)->String().Data());
+            continue;
+         }
+
+         SetupAxes(*hs);
+
+         // Perform projection.
+         TObject* h = nullptr;
+
+         if (ndim == 1) {
+            h = static_cast<TObject*>(hs->Projection(dim[0], option));
```

Comment:
We usually don't use `static_cast` for cast-to-base... I don't think it adds much either. Could you remove these (i.e. also below)?